frame/transaction-payment/asset-conversion-tx-payment/src/payment.rs

```diff
@@ -169,6 +174,15 @@ where
 			// If this fails, the account might have dropped below the existential balance or there
 			// is not enough liquidity left in the pool. In that case we don't throw an error and
 			// the account will keep the native currency.
+
+			let amount_out_min = if let Some(max_fee) = max_fee {
+				let worst_native_to_asset_rate = T::HigherPrecisionBalance::from(max_fee) /
+					T::HigherPrecisionBalance::from(fee_paid);
+				Some(T::HigherPrecisionBalance::from(swap_back) * worst_native_to_asset_rate)
+			} else {
+				None
+			};
```

Comment:
This is the calculation to review carefully. Everything else is pretty tame.

Reply:
I don't understand why this is needed post-dispatch. The only thing that can happen now is a refund. If the pre-dispatch `withdraw_fee` respects the maximum taken, there's no way it gets worse in post. Shouldn't we just take back whatever we can get?

Reply:
Good point - only the executed transaction could change it from the rate it was at the start of the transaction, so it's safe to have None for the post-dispatch.
Let me simplify it.